enhancements/agent-installer/simplified-registry-less-cluster-operations.md

```diff
@@ -0,0 +1,426 @@
+
+### Installation
+
+#### Bootstrap phase
+
+Based on the same [Agent-based Installer](https://github.com/openshift/enhancements/blob/master/enhancements/agent-installer/automated-workflow-for-agent-based-installer.md)(ABI)
+approach, the installation will be kicked off into an ephemeral environment
+fully allocated in memory, as soon as each node is booted using the extended
+ISO (bootstrap phase).
+In addition to the usual ABI services configured to orchestrate the
+installation, a local registry service will be created on each node, and the
+content of the release payload from the mounted ISO will be used as its
+backend source.
+During this early phase, the main responsibility of the local registry will
+be to serve any image pull request originating from the local services running
+on the same node.
+
+#### Installation phase (after the first reboot)
+
+During the installation of each control plane node, the contents of the OCP
+release payload will be copied from the ISO to the node's disk (in the format
+of the registry's 'filesystem' backend) after the OS disk writing step - so
+that it remains available not only after the first reboot, but also once the
+installation is complete. Also the registry container image itself will be
+copied in the node additional image store for podman.
+[Assisted Service](https://github.com/openshift/assisted-service/)
+and [Assisted Installer](https://github.com/openshift/assisted-installer)
+will be enhanced accordingly to take care of this step, and to ensure it
+will be properly monitored during the installation process.
+
+A new registry service will be injected via MCO to support providing the
+release content stored on disk after the reboot (since the previous ephemeral
+environment will not be available anymore).
+
+To allow pulling images from the cluster, an ImageDigestMirrorSet
+(IDMS) policy will be added to the cluster manifests - using the `api-int` 
+DNS entry - so that the set of registries running within
+the control plane nodes will become the nominal pullspec of the current release
+payload itself.
+
+Finally, a new `InternalReleaseImage` custom resource, managed by MCO, will
+also be added as a cluster manifest to keep track and manage the currently
+stored internal release image.
+
+### Upgrade
+
+Upgrading an existing lower version cluster will be performed via the same
+extended ISO used for managing the installation. As soon as the user will
+attach the extended ISO to one of the control planes a temporary registry
+will be created to serve its content.
+At this point the user could edit the existing IRI resource to add the new
+release to trigger the copy process (from the temporary registry
+to each local control plane node registry), managed by the Machine Config
+Daemon (MCD). Once the copy has been completed on all the control plane nodes,
+the user could start the usual offline upgrade process via the `oc adm upgrade
+--allow-explicit-upgrade --to-image=<release-image>`.
```

Comment:
What is the suggested approach to extract the relevant 'release-image' from the ISO?
Or maybe since the ISO is constructed for a specific release image anyway, we can keep it transparent for users? E.g. apply a MachineConfig that starts the upgrade (and detects the right version internally)

Reply:
I've lost track of the specific validations the CVO does but I believe the way it's intended to work today is if you provide a symbolic version it has to be on the graph. As soon as you provide a specific pull spec (--allow-explicit-upgrade) the CVO assumes you know what you're doing, not that that's ever been documented explicitly. There's a lot of blurred lines between CLI side guards and what the API accepts.

@wking can you check me here? what protections are in place here and what do you think would be best way to keep them from making a mistake by potentially selecting the wrong mirrored image when expecting to upgrade? Maybe it's time to move all the client side protections into the ClusterVersion CR and CVO via a new interface?

Reply:
It would be great if we had symbolic versions. What I think we'd need to do that is a mapping of tags to sha digests that is cryptographically signed. If that existed it would be fairly easy to build into the ISO, but the infrastructure to set it up if it doesn't would be potentially daunting. Do you know of any existing artifacts that could act as building blocks here? A brief reading of sigstore docs suggests that it won't help.

Reply:
FWIW: Before OSUS (and oc-mirror) were around, when performing OCP4 upgrades on disconnected environments, they had to be done with `--allow-explicit-upgrade` and that did have the risks you are pointing out here of not properly obeying the graph. In that era, the message used to be that you had to be careful to follow the graph on your own or you risked to not be properly supported.

However, I'd prefer to have safeguard mechanisms like the ones you are considering.

Reply:
Arguably there's a high probability that most disconnected upgrades still follow this path and don't use OSUS, and that we should therefore punt this to a future enhancement rather than block on it now. (I agree that I'd _prefer_ to have it, all things being equal.)